server/mediaserver.go

```diff
@@ -283,7 +289,8 @@ func gotRTMPStreamHandler(s *LivepeerServer) func(url *url.URL, rtmpStrm stream.
 		glog.V(common.SHORT).Infof("\n\nhlsStrmID: %v\n\n", hlsStrmID)
 
 		//Create Transcode Job Onchain
-		go s.startSession(cxn)
+		go s.startSessionListener(cxn)
+		cxn.needOrch <- struct{}{}
```

Comment:
What ensures that the select statement in `s.startSessionListener` is actually running/listening when line 293 gets called? Trying to again figure out concurrency in go!

Reply:
> What ensures that the select statement in s.startSessionListener is actually running/listening when line 293 gets called

Nothing actually. If the RTMP stream stops, `eof` is called and the select statement exits, we'll still be feeding `cxn.needOrch` for each segment that is in-flight at the time of `eof`.

This seems OK. We could `close` the needOrch chanel instead of using an `eof` chan, but that panics (crashes) if we ever inadvertently write to a closed channel. It seems safer to use an `eof` chan and accept the small hit of writing to a channel that will never be drained. The GC ensures that we eventually reclaim any empty `struct{}`s that may have been buffered in the meantime.

Reply:
Interesting! Didn't know about the crash if we write to a closed channel. Agreed with your thinking here. I'm wondering though if `s.startSessionListener` will be listening the first time `cxn.needOrch <- struct{}{}` gets called on line 293? i.e. do we know with certainty that the`for{}` statement on line 365 would get executed before line 293 every time?

Reply:
Ah I see. If the scheduling is such that `cxn.needOrch` is written to before `startSessionListener` completes the `for/select` setup, then the `needOrch` chan will have its contents buffered until the `for/select` drains it (or `needOrch` is GC'd at stream exit).

Reply:
Ooooo I see! That makes sense. +1 for the creators of Golang

Reply:
Heh. I'd argue that the opaqueness of most of channel semantics is a drawback here -- someone made a choice for a lot of these behaviors; they have to be learned and can't be derived or reasoned about as with traditional concurrency primitives. That might be justified if chans were implemented as a library atop the base, but it's somewhat questionable as a general-purpose, first-class language construct.